### PR TITLE
Stylo uplifts 3

### PR DESCRIPTION
**Pull request by bholley**
[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9852)

### CODE REVIEW

**Comment by nox** [comment]
> I wanted to first just update the existing bindings to match the state of the world in gecko.

How did you do that?

**Comment by nox** [comment]
@bholley I don't think the first step to making bindings generation a pleasant task is to commit undocumented scripts to the repository. The first step is to document the whole process, why we need a Clang fork for example. I'm not sure we need one anymore.

File: components/style/selector_matching.rs [hunk 1]
```diff
@@ -146,7 +146,7 @@ impl<Impl: SelectorImplExt> Stylist<Impl> {
         }
 
         if self.quirks_mode {
-            self.add_stylesheet(&Impl::get_quirks_mode_stylesheet());
+            Impl::get_quirks_mode_stylesheet().map(|s| self.add_stylesheet(s));
```
Comment (ecoal95) [inline components/style/selector_matching.rs]: Not too relevant, but I prefer the `if let Some(..)` idiom for things that have side effects and which return result you don't need.

**Comment by bholley** [comment]
> @bholley I don't think the first step to making bindings generation a pleasant task is to commit undocumented scripts to the repository.

We may disagree on what the first step is, but this is clearly an improvement over the status quo (where there was no guidance on how to regenerate the bindings at all), so let's not make perfect the enemy of good.

PRs to improve any of this are, of course, welcome.

> The first step is to document the whole process, why we need a Clang fork for example. 

I'm just automating the instructions in https://github.com/michaelwu/rust-bindgen/blob/sm-hacks/README.md

> I'm not sure we need one anymore.

It largely depends on what the user has installed, I think. PRs welcome to detect the right thing, skip the build in that case, and put the it in LIBCLANG_PATH.

**Comment by nox** [comment]
> We may disagree on what the first step is, but this is clearly an improvement over the status quo (where there was no guidance on how to regenerate the bindings at all), so let's not make perfect the enemy of good.

I stand corrected.

**Comment by SimonSapin** [comment]
r=me with @ecoal95’s nit addressed

@bors-servo delegate+

**Pull request merged**